include/clap/factory/draft/preset-discovery.h

```diff
@@ -0,0 +1,315 @@
+/*
+   Preset Discovery API.
+
+   Preset Discovery enables a plug-in host to identify where presets are found, what
+   extensions they have, which plug-ins they apply to, and other metadata associated with the
+   presets so that they can be indexed and searched for quickly within the plug-in host's browser.
+
+   This has a number of advantages for the user:
+   - it allows them to browse for presets from one central location in a consistent way
+   - the user can browse for presets without having to commit to a particular plug-in first
+
+   The API works as follow to index presets and presets metadata:
+   1. clap_plugin_entry.get_factory(CLAP_PRESET_DISCOVERY_FACTORY_ID)
+   2. clap_preset_discovery_factory_t.create(...)
+   3. clap_preset_discovery_provider.init() (only necessary the first time, declarations
+   can be cached)
+        `-> clap_preset_discovery_indexer.declare_filetype()
+        `-> clap_preset_discovery_indexer.declare_location()
+        `-> clap_preset_discovery_indexer.declare_collection() (optional)
+        `-> clap_preset_discorery_indexer.set_invalidation_watch_file() (optional)
+   4. crawl the given locations and monitor file system changes
+        `-> clap_preset_discovery_indexer.get_metadata() for each presets files
```

Comment:
Bit of a radical suggestion, but would adding a `clap_plugin_` and `clap_host_` prefix to `clap_preset_discovery_provider`, and `clap_preset_discovery_indexer`, and `clap_preset_discovery_metadata_receiver` introduce too much clutter? I always forget which side uses which, and I'm sure anyone else first looking at this extension will run into the same issue. Consistent naming conventions might make this a bit simpler to grasp.

Also, this reads clap\_preset\_**discorery**\_indexer right now. :wink:

Reply:
> Bit of a radical suggestion, but would adding a `clap_plugin_` and `clap_host_` prefix to `clap_preset_discovery_provider`, and `clap_preset_discovery_indexer`, and `clap_preset_discovery_metadata_receiver` introduce too much clutter?

I also would vote against host/plugin prefixes for factory methods. As extensions are per plugin (instance), the prefixes make sense for extension methods. Factories are per .clap file and don't necessarily apply to a single plugin. (Also it isn't given that there already is a clap host when calling a factory method. Only when the plugin instances are actually created, the host is needed.)

Reply:
@robbert-vdh Yeah, I think you are right (as well?). When talking about host/plugin I meant the `clap_host_t`/`clap_plugin_t` structs. I just think that it's easy to confuse both host/plugin levels ATM. And those prefixes would confuse it even more.